File: src-tauri/src/analysis/context_tracker/injections.rs
use serde::{Deserialize, Serialize};

/// Context injection types — tagged union mirroring the TS discriminated union.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "category", rename_all = "kebab-case")]
pub enum ContextInjection {
    #[serde(rename = "claude-md")]
    ClaudeMd(ClaudeMdContextInjection),
    #[serde(rename = "mentioned-file")]
    MentionedFile(MentionedFileInjection),
    #[serde(rename = "tool-output")]
    ToolOutput(ToolOutputInjection),
    #[serde(rename = "thinking-text")]
    ThinkingText(ThinkingTextInjection),
    #[serde(rename = "task-coordination")]
    TaskCoordination(TaskCoordinationInjection),
    #[serde(rename = "user-message")]
    UserMessage(UserMessageInjection),
}

impl ContextInjection {
    pub fn estimated_tokens(&self) -> u64 {
        match self {
            Self::ClaudeMd(i) => i.estimated_tokens,
            Self::MentionedFile(i) => i.estimated_tokens,
            Self::ToolOutput(i) => i.estimated_tokens,
            Self::ThinkingText(i) => i.estimated_tokens,
            Self::TaskCoordination(i) => i.estimated_tokens,
            Self::UserMessage(i) => i.estimated_tokens,
        }
    }

    pub fn path(&self) -> Option<&str> {
        match self {
            Self::ClaudeMd(i) => Some(&i.path),
            Self::MentionedFile(i) => Some(&i.path),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClaudeMdContextInjection {
    pub id: String,
    pub path: String,
    pub source: String,
    pub display_name: String,
    pub is_global: bool,
    pub estimated_tokens: u64,
    pub first_seen_in_group: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MentionedFileInjection {
    pub id: String,
    pub path: String,
    pub display_name: String,
    pub estimated_tokens: u64,
    pub first_seen_turn_index: u32,
    pub first_seen_in_group: String,
    #[serde(default = "default_true")]
    pub exists: bool,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolOutputInjection {
    pub id: String,
    pub turn_index: u32,
    pub ai_group_id: String,
    pub estimated_tokens: u64,
    pub tool_count: u32,
    pub tool_breakdown: Vec<ToolTokenBreakdown>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolTokenBreakdown {
    pub tool_name: String,
    pub token_count: u64,
    pub is_error: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_use_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThinkingTextInjection {
    pub id: String,
    pub turn_index: u32,
    pub ai_group_id: String,
    pub estimated_tokens: u64,
    pub breakdown: Vec<ThinkingTextBreakdown>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThinkingTextBreakdown {
    #[serde(rename = "type")]
    pub item_type: String,
    pub token_count: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskCoordinationInjection {
    pub id: String,
    pub turn_index: u32,
    pub ai_group_id: String,
    pub estimated_tokens: u64,
    pub breakdown: Vec<TaskCoordinationBreakdown>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskCoordinationBreakdown {
    #[serde(rename = "type")]
    pub item_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_name: Option<String>,
    pub token_count: u64,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UserMessageInjection {
    pub id: String,
    pub turn_index: u32,
    pub ai_group_id: String,
    pub estimated_tokens: u64,
    pub text_preview: String,
}

// Token category totals
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokensByCategory {
    pub claude_md: u64,
    pub mentioned_files: u64,
    pub tool_outputs: u64,
    pub thinking_text: u64,
    pub task_coordination: u64,
    pub user_messages: u64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewCountsByCategory {
    pub claude_md: u32,
    pub mentioned_files: u32,
    pub tool_outputs: u32,
    pub thinking_text: u32,
    pub task_coordination: u32,
    pub user_messages: u32,
}

// ID generation functions

pub fn generate_mentioned_file_id(path: &str) -> String {
    let mut hash: i32 = 0;
    for ch in path.chars() {
        hash = hash.wrapping_shl(5).wrapping_sub(hash).wrapping_add(ch as i32);
    }
    format!("mf-{:x}", hash.unsigned_abs())
}

pub fn generate_tool_output_id(turn_index: u32) -> String {
    format!("tool-output-ai-{}", turn_index)
}

pub fn generate_thinking_text_id(turn_index: u32) -> String {
    format!("thinking-text-ai-{}", turn_index)
}

pub fn generate_task_coordination_id(turn_index: u32) -> String {
    format!("task-coord-ai-{}", turn_index)
}

pub fn generate_user_message_id(turn_index: u32) -> String {
    format!("user-msg-ai-{}", turn_index)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mentioned_file_id_deterministic() {
        let id1 = generate_mentioned_file_id("/Users/test/file.ts");
        let id2 = generate_mentioned_file_id("/Users/test/file.ts");
        assert_eq!(id1, id2);
        assert!(id1.starts_with("mf-"));
    }

    #[test]
    fn different_paths_different_ids() {
        let id1 = generate_mentioned_file_id("/a/b.ts");
        let id2 = generate_mentioned_file_id("/c/d.ts");
        assert_ne!(id1, id2);
    }

    #[test]
    fn tool_output_id_format() {
        assert_eq!(generate_tool_output_id(5), "tool-output-ai-5");
    }

    #[test]
    fn thinking_text_id_format() {
        assert_eq!(generate_thinking_text_id(3), "thinking-text-ai-3");
    }

    #[test]
    fn task_coordination_id_format() {
        assert_eq!(generate_task_coordination_id(7), "task-coord-ai-7");
    }

    #[test]
    fn user_message_id_format() {
        assert_eq!(generate_user_message_id(0), "user-msg-ai-0");
    }

    #[test]
    fn context_injection_serde_roundtrip() {
        let injection = ContextInjection::UserMessage(UserMessageInjection {
            id: "user-msg-ai-0".to_string(),
            turn_index: 0,
            ai_group_id: "ai-0".to_string(),
            estimated_tokens: 50,
            text_preview: "Hello world".to_string(),
        });
        let json = serde_json::to_string(&injection).unwrap();
        assert!(json.contains("\"category\":\"user-message\""));
        let deserialized: ContextInjection = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.estimated_tokens(), 50);
    }
}

File: src/renderer/utils/contextTracker/injections.ts
import { estimateTokens } from '@shared/utils/tokenFormatting';

import type { ClaudeMdInjection } from '../../types/claudeMd';
import type {
  ClaudeMdContextInjection,
  MentionedFileInjection,
  TaskCoordinationBreakdown,
  TaskCoordinationInjection,
  ThinkingTextBreakdown,
  ThinkingTextInjection,
  ToolOutputInjection,
  ToolTokenBreakdown,
  UserMessageInjection,
} from '../../types/contextInjection';
import type { AIGroupDisplayItem, LinkedToolItem, UserGroup } from '../../types/groups';

export const CATEGORY_MENTIONED_FILE = 'mentioned-file' as const;

export const TASK_COORDINATION_TOOL_NAMES = new Set([
  'SendMessage',
  'TeamCreate',
  'TeamDelete',
  'TaskCreate',
  'TaskUpdate',
  'TaskList',
  'TaskGet',
]);

function generateMentionedFileId(path: string): string {
  let hash = 0;
  for (let i = 0; i < path.length; i++) {
    const char = path.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash;
  }
  const positiveHash = Math.abs(hash).toString(16);
  return `mf-${positiveHash}`;
}

function generateToolOutputId(turnIndex: number): string {
  return `tool-output-ai-${turnIndex}`;
}

function generateThinkingTextId(turnIndex: number): string {
  return `thinking-text-ai-${turnIndex}`;
}

function generateTaskCoordinationId(turnIndex: number): string {
  return `task-coord-ai-${turnIndex}`;
}

function generateUserMessageId(turnIndex: number): string {
  return `user-msg-ai-${turnIndex}`;
}

export function wrapClaudeMdInjection(injection: ClaudeMdInjection): ClaudeMdContextInjection {
  return {
    ...injection,
    category: 'claude-md' as const,
  };
}

interface CreateMentionedFileInjectionParams {
  path: string;
  displayName: string;
  estimatedTokens: number;
  turnIndex: number;
  aiGroupId: string;
  exists?: boolean;
}

export function createMentionedFileInjection(
  params: CreateMentionedFileInjectionParams
): MentionedFileInjection {
  return {
    id: generateMentionedFileId(params.path),
    category: CATEGORY_MENTIONED_FILE,
    path: params.path,
    displayName: params.displayName,
    estimatedTokens: params.estimatedTokens,
    firstSeenTurnIndex: params.turnIndex,
    firstSeenInGroup: params.aiGroupId,
    exists: params.exists ?? true,
  };
}

export function aggregateToolOutputs(
  linkedTools: Map<string, LinkedToolItem>,
  turnIndex: number,
  aiGroupId: string,
  displayItems?: AIGroupDisplayItem[]
): ToolOutputInjection | null {
  const toolBreakdown: ToolTokenBreakdown[] = [];
  let totalTokens = 0;

  for (const linkedTool of linkedTools.values()) {
    if (TASK_COORDINATION_TOOL_NAMES.has(linkedTool.name)) {
      continue;
    }

    const callTokens = linkedTool.callTokens ?? 0;
    const resultTokens = linkedTool.result?.tokenCount ?? 0;
    const skillTokens = linkedTool.skillInstructionsTokenCount ?? 0;
    const toolTokenCount = callTokens + resultTokens + skillTokens;

    if (toolTokenCount > 0) {
      const displayName = linkedTool.name === 'Task' ? 'Task (Subagent)' : linkedTool.name;
      toolBreakdown.push({
        toolName: displayName,
        tokenCount: toolTokenCount,
        isError: linkedTool.result?.isError ?? false,
        toolUseId: linkedTool.id,
      });
      totalTokens += toolTokenCount;
    }
  }

  if (displayItems) {
    for (const item of displayItems) {
      if (item.type === 'slash' && item.slash.instructionsTokenCount) {
        toolBreakdown.push({
          toolName: `/${item.slash.name}`,
          tokenCount: item.slash.instructionsTokenCount,
          isError: false,
        });
        totalTokens += item.slash.instructionsTokenCount;
      }
    }
  }

  if (totalTokens === 0) {
    return null;
  }

  return {
    id: generateToolOutputId(turnIndex),
    category: 'tool-output',
    turnIndex,
    aiGroupId,
    estimatedTokens: totalTokens,
    toolCount: toolBreakdown.length,
    toolBreakdown,
  };
}

export function aggregateTaskCoordination(
  linkedTools: Map<string, LinkedToolItem>,
  turnIndex: number,
  aiGroupId: string,
  displayItems?: AIGroupDisplayItem[]
): TaskCoordinationInjection | null {
  const breakdown: TaskCoordinationBreakdown[] = [];
  let totalTokens = 0;

  for (const linkedTool of linkedTools.values()) {
    if (!TASK_COORDINATION_TOOL_NAMES.has(linkedTool.name)) {
      continue;
    }

    const callTokens = linkedTool.callTokens ?? 0;
    const resultTokens = linkedTool.result?.tokenCount ?? 0;
    const skillTokens = linkedTool.skillInstructionsTokenCount ?? 0;
    const toolTokenCount = callTokens + resultTokens + skillTokens;

    if (toolTokenCount > 0) {
      let label = linkedTool.name;
      if (linkedTool.name === 'SendMessage' && linkedTool.input) {
        const recipient = linkedTool.input.recipient as string | undefined;
        if (recipient) {
          label = `SendMessage → ${recipient}`;
        }
      }

      breakdown.push({
        type: linkedTool.name === 'SendMessage' ? 'send-message' : 'task-tool',
        toolName: linkedTool.name,
        tokenCount: toolTokenCount,
        label,
      });
      totalTokens += toolTokenCount;
    }
  }

  if (displayItems) {
    for (const item of displayItems) {
      if (item.type === 'teammate_message' && item.teammateMessage.tokenCount) {
        breakdown.push({
          type: 'teammate-message',
          tokenCount: item.teammateMessage.tokenCount,
          label: item.teammateMessage.teammateId,
        });
        totalTokens += item.teammateMessage.tokenCount;
      }
    }
  }

  if (totalTokens === 0) {
    return null;
  }

  return {
    id: generateTaskCoordinationId(turnIndex),
    category: 'task-coordination',
    turnIndex,
    aiGroupId,
    estimatedTokens: totalTokens,
    breakdown,
  };
}

export function createUserMessageInjection(
  userGroup: UserGroup,
  turnIndex: number,
  aiGroupId: string
): UserMessageInjection | null {
  const text = userGroup.content.rawText ?? userGroup.content.text ?? '';
  if (!text) return null;

  const tokens = estimateTokens(text);
  if (tokens === 0) return null;

  const textPreview = text.length > 80 ? text.slice(0, 80) + '…' : text;

  return {
    id: generateUserMessageId(turnIndex),
    category: 'user-message',
    turnIndex,
    aiGroupId,
    estimatedTokens: tokens,
    textPreview,
  };
}

export function aggregateThinkingText(
  displayItems: AIGroupDisplayItem[],
  turnIndex: number,
  aiGroupId: string
): ThinkingTextInjection | null {
  const breakdown: ThinkingTextBreakdown[] = [];
  let totalTokens = 0;
  let thinkingTokens = 0;
  let textTokens = 0;

  for (const item of displayItems) {
    if (item.type === 'thinking' && item.tokenCount && item.tokenCount > 0) {
      thinkingTokens += item.tokenCount;
      totalTokens += item.tokenCount;
    } else if (item.type === 'output' && item.tokenCount && item.tokenCount > 0) {
      textTokens += item.tokenCount;
      totalTokens += item.tokenCount;
    }
  }

  if (thinkingTokens > 0) {
    breakdown.push({ type: 'thinking', tokenCount: thinkingTokens });
  }
  if (textTokens > 0) {
    breakdown.push({ type: 'text', tokenCount: textTokens });
  }

  if (totalTokens === 0) {
    return null;
  }

  return {
    id: generateThinkingTextId(turnIndex),
    category: 'thinking-text',
    turnIndex,
    aiGroupId,
    estimatedTokens: totalTokens,
    breakdown,
  };
}
